### PR TITLE
feat: Add Converse component and improve output extraction

### PR DESCRIPTION
**Pull request by Cristhianzl**
This pull request adds comprehensive tests for field name conversion and output extraction utilities, and updates async handling in agent test scripts for consistency. The main changes include extensive new test coverage for field name conversion between camelCase and snake_case, thorough tests for the `_extract_output_text` function (including edge cases and backward compatibility), and updates to use the `run_until_complete` helper for asynchronous toolkit initialization in agent-related tests.

**Testing improvements:**

* Added a comprehensive test class `TestFieldNameConversion` to `test_mcp_util.py` to cover all scenarios of camelCase to snake_case conversion, including edge cases, integration with Pydantic schemas, and value preservation. Also added `TestToolExecutionWithFieldConversion` to verify that tool execution functions correctly handle field name conversion in both async and sync contexts.
* Added a large suite of tests for the `_extract_output_text` function in `test_agent_events.py`, covering various input formats, edge cases, backward compatibility, and specific scenarios related to ChatBedrockConverse and streaming responses.

**Async handling consistency:**

* Updated agent test scripts (`simple_agent.py` and `test_simple_agent_in_lfx_run.py`) to use the `run_until_complete` helper from `lfx.utils.async_helpers` for initializing toolkits, ensuring consistent async execution patterns across tests. [[1]](diffhunk://#diff-e246f55cbaa4c296efba4adfb5fc0f0846667ab6700843ecc789580859384719L17-R24) [[2]](diffhunk://#diff-e246f55cbaa4c296efba4adfb5fc0f0846667ab6700843ecc789580859384719L35-R35) [[3]](diffhunk://#diff-2de3ee12e172b9d07123c6e693c19c3356215a0c4141416ff47080bd6c05ca47L55-R55) [[4]](diffhunk://#diff-2de3ee12e172b9d07123c6e693c19c3356215a0c4141416ff47080bd6c05ca47R149) [[5]](diffhunk://#diff-2de3ee12e172b9d07123c6e693c19c3356215a0c4141416ff47080bd6c05ca47L164-R165)

**Test imports and dependencies:**

* Added missing import of `_extract_output_text` in `test_agent_events.py` to enable direct testing of this utility.

### CODE REVIEW

**Comment by edwinjosechittilappilly** [comment]
> Is the current Bedrock component useful to keep it or should we just make it use the Converse version?

I would suggest to deprecate the old one since it might be using the not supported LMbackend of AWS

**Comment by edwinjosechittilappilly** [comment]
@Cristhianzl  Please take a look at the Conflicts also This being a non core component may be we can deprecate the old AWS component.

File: src/lfx/src/lfx/base/mcp/util.py [hunk 1]
```diff
@@ -189,6 +189,37 @@ def sanitize_mcp_name(name: str, max_length: int = 46) -> str:
...
+            # Try converting camelCase to snake_case
+            snake_key = _camel_to_snake(key)
+            if snake_key in schema_fields:
+                converted_args[snake_key] = value
+            else:
+                # If neither the original nor converted key exists, keep original
+                # The validation will catch this error
+                converted_args[key] = value
+
+    return converted_args
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py:221]: You may look into libraries like https://github.com/sanders41/camel-converter

File: src/lfx/src/lfx/base/mcp/util.py [hunk 2]
```diff
@@ -202,10 +233,22 @@ async def tool_coroutine(*args, **kwargs):
             provided_args[field_names[i]] = arg
         # Merge in keyword arguments
         provided_args.update(kwargs)
+        # Convert camelCase field names to snake_case if needed
+        provided_args = _convert_field_names(provided_args, arg_schema)
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py]: would make this a more descriptive method name so we don't have to add the comment

File: src/lfx/src/lfx/base/mcp/util.py [hunk 3]
```diff
@@ -202,10 +233,22 @@ async def tool_coroutine(*args, **kwargs):
...
             validated = arg_schema.model_validate(provided_args)
         except Exception as e:
+            # Check if this is a case where the tool was called with no arguments
+            if not provided_args and hasattr(arg_schema, "model_fields"):
+                required_fields = [name for name, field in arg_schema.model_fields.items() if field.is_required()]
+                if required_fields:
+                    msg = (
+                        f"Tool '{tool_name}' requires arguments but none were provided. "
+                        f"Required fields: {', '.join(required_fields)}. "
+                        f"Please check that the LLM is properly calling the tool with arguments."
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py]: Is this something the User can influence? i.e. will they know that they can improve their prompt or tool description to handle this case?

File: src/lfx/src/lfx/base/mcp/util.py [hunk 4]
```diff
@@ -230,9 +273,21 @@ def tool_func(*args, **kwargs):
...
+                    )
+                    raise ValueError(msg) from e
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py]: Would separate to function if repeated

File: src/lfx/src/lfx/base/mcp/util.py [hunk 5]
```diff
@@ -1195,6 +1252,8 @@ async def run_tool(self, tool_name: str, arguments: dict[str, Any]) -> Any:
         Raises:
             ValueError: If session is not initialized or tool execution fails
         """
+        import asyncio
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py]: may as well move them to the top

File: src/lfx/src/lfx/base/mcp/util.py [hunk 6]
```diff
@@ -1378,7 +1434,69 @@ async def update_tools(
                 logger.warning(f"Could not create schema for tool '{tool.name}' from server '{server_name}'")
                 continue
 
-            tool_obj = StructuredTool(
+            # Create a custom StructuredTool that bypasses schema validation
+            class MCPStructuredTool(StructuredTool):
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py:1429]: why do we need to bypass schema validation?

File: src/lfx/src/lfx/base/mcp/util.py [hunk 7]
```diff
@@ -1378,7 +1434,69 @@ async def update_tools(
...
+
+                    # Call the parent run method with converted parameters
+                    return super().run(converted_input, config=config, **kwargs)
+
+                async def arun(self, tool_input: str | dict, config=None, **kwargs):
+                    """Override the main arun method to handle parameter conversion before validation."""
+                    # Parse tool_input if it's a string
+                    if isinstance(tool_input, str):
+                        try:
+                            import json
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py]: move these to top

File: src/lfx/src/lfx/base/mcp/util.py [hunk 8]
```diff
@@ -1378,7 +1434,69 @@ async def update_tools(
                 logger.warning(f"Could not create schema for tool '{tool.name}' from server '{server_name}'")
                 continue
 
-            tool_obj = StructuredTool(
+            # Create a custom StructuredTool that bypasses schema validation
+            class MCPStructuredTool(StructuredTool):
+                def run(self, tool_input: str | dict, config=None, **kwargs):
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/base/mcp/util.py:1430]: do you have tests for these inner methods?

File: src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py [hunk 9]
```diff
@@ -14,6 +14,7 @@ class AmazonBedrockComponent(LCModelComponent):
     )
     icon = "Amazon"
     name = "AmazonBedrockModel"
+    legacy = True
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py:17]: Can we add the "recommended component that replaced this"

File: src/lfx/src/lfx/schema/json_schema.py [hunk 10]
```diff
@@ -2,13 +2,19 @@
...
 
 from lfx.log.logger import logger
 
 NULLABLE_TYPE_LENGTH = 2  # Number of types in a nullable union (the type itself + null)
 
 
+def _snake_to_camel(name: str) -> str:
+    """Convert snake_case to camelCase."""
+    components = name.split("_")
+    return components[0] + "".join(word.capitalize() for word in components[1:])
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/schema/json_schema.py]: Is there any reason we'd want to preserve leading or trailing underscores? e.g. `_my_variable -> _myVariable`?

File: src/lfx/src/lfx/schema/json_schema.py [hunk 11]
```diff
@@ -118,7 +124,14 @@ def _build_model(name: str, subschema: dict[str, Any]) -> type[BaseModel]:
             else:
                 default = ...  # required by Pydantic
 
-            fields[prop_name] = (py_type, Field(default, description=prop_schema.get("description")))
+            # Add alias for camelCase if field name is snake_case
+            field_kwargs = {"description": prop_schema.get("description")}
+            if "_" in prop_name:
+                camel_case_name = _snake_to_camel(prop_name)
+                if camel_case_name != prop_name:  # Only add alias if it's different
+                    field_kwargs["validation_alias"] = AliasChoices(prop_name, camel_case_name)
```
Comment (jordanrfrazier) [inline src/lfx/src/lfx/schema/json_schema.py:156]: I feel like I need an explanation on why this is necessary. Is that already in the mcp comments or description somewhere?

File: src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py (hunk 9 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py:17]: sure.

File: src/lfx/src/lfx/base/mcp/util.py (hunk 8 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/base/mcp/util.py:1430]: added the tests.

File: src/lfx/src/lfx/schema/json_schema.py (hunk 10 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/schema/json_schema.py]: There isn’t a specific reason, but I think it’s interesting to include this.
Thank you.

File: src/lfx/src/lfx/base/mcp/util.py (hunk 3 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/base/mcp/util.py]: Actually, the users don’t have access to or the ability to modify that part of the errors

File: src/lfx/src/lfx/schema/json_schema.py (hunk 11 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/schema/json_schema.py:156]: When Langflow flows are exposed as MCP tools, they need to handle input from different MCP clients, which may send parameters in camelCase.
However, Langflow’s internal schemas expect snake_case field names.

Without alias support:
A client sending {"weatherMain": "Rain", "topN": 5} would fail validation
Only {"weather_main": "Rain", "top_n": 5} would be accepted

Solution:
Alias functionality allows a single Pydantic model to accept both naming conventions seamlessly, ensuring compatibility with all clients.

File: src/backend/tests/data/simple_agent.py [hunk 12]
```diff
@@ -14,16 +14,18 @@
...
 from pathlib import Path
 
 from lfx.graph import Graph
 from lfx.log.logger import LogConfig
+from lfx.utils.async_helpers import run_until_complete
 
 # Using the new flattened component access
 from lfx import components as cp
 
+# Using the new flattened component access
```
Comment (jordanrfrazier) [inline src/backend/tests/data/simple_agent.py]: what is this?

File: src/backend/tests/unit/base/mcp/test_mcp_util.py [hunk 13]
```diff
@@ -251,6 +251,446 @@ async def test_sse_client_header_storage(self):
...
+
+        # Mock asyncio.get_event_loop
+        mock_loop = MagicMock()
+        mock_loop.run_until_complete = MagicMock(return_value="sync_result")
+
+        with patch("asyncio.get_event_loop", return_value=mock_loop):
+            # Test with camelCase fields
+            result = tool_func(userName="testuser", maxResults=10)
+
+            assert result == "sync_result"
```
Comment (jordanrfrazier) [inline src/backend/tests/unit/base/mcp/test_mcp_util.py:690]: Are these mocks still testing what we want them to?

File: src/lfx/src/lfx/components/amazon/amazon_bedrock_converse.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
+from langflow.base.models.aws_constants import AWS_REGIONS, AWS_MODEL_IDs
+from langflow.base.models.model import LCModelComponent
+from langflow.field_typing import LanguageModel
+from langflow.inputs.inputs import BoolInput, FloatInput, IntInput, MessageTextInput, SecretStrInput
+from langflow.io import DictInput, DropdownInput
+
```
[the remark this answers is not in the document]
Comment (jordanrfrazier) [reply src/lfx/src/lfx/components/amazon/amazon_bedrock_converse.py]: Is this true? I see you resolved but just checking since the imports could technically be moved to lfx

File: src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py (hunk 9 above)
Comment (jordanrfrazier) [reply src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py:17]: There's a flag called `replacement`. 

e.g.
```
    replacement = "amazon.AmazonConverseWhateverTheCorrectNameIs"
```

File: src/lfx/src/lfx/schema/json_schema.py (hunk 11 above)
Comment (jordanrfrazier) [reply src/lfx/src/lfx/schema/json_schema.py:156]: Awesome. Would mention an abbreviated version of this somewhere wherever this is being done in MCP

**Review by jordanrfrazier — APPROVED** [review]

File: src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py (hunk 9 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/components/amazon/amazon_bedrock_model.py:17]: oh nice, I didn't know about this flag.

File: src/lfx/src/lfx/schema/json_schema.py (hunk 11 above)
Comment (Cristhianzl) [reply src/lfx/src/lfx/schema/json_schema.py:156]: yeah, I can send you a SSE URL where you can find tools in camel case

File: src/backend/tests/data/simple_agent.py (hunk 12 above)
Comment (Cristhianzl) [reply src/backend/tests/data/simple_agent.py]: just merge issue with the comments.

File: src/backend/tests/unit/base/mcp/test_mcp_util.py (hunk 13 above)
Comment (Cristhianzl) [reply src/backend/tests/unit/base/mcp/test_mcp_util.py:690]: yes, just various scenarios testing the in/out of the function xD
